fix: Numerical URL targets cause site hang

chenc041 opened the pull request:
if location hash is start of numerical, it will cause useScrollRestoration hook throw an error and infinity loop 

change file  [package/remix-react/scroll-restoration.tsx](https://github.com/remix-run/remix/blob/main/packages/remix-react/scroll-restoration.tsx#L101)

issue num [#935](https://github.com/remix-run/remix/issues/935)

reference [Rule sets, declaration blocks, and selectors](https://www.w3.org/TR/CSS2/syndata.html#characters)

tascord reviewed (commented):
Looks good!
Not sure about ettiquite in terms of the `.idea` directory in this project so no comment there.
I'll have a look at minifying the scroll-restoration fix or maybe making it a bit more understandable w/ comments.

tascord reviewed (approved):
👍

chenc041 commented:
> 👍

😄

MichaelDeBoey commented on .gitignore:
```diff
@@ -11,3 +11,4 @@ yarn-error.log
 /fixtures/my-remix-app
 
 .eslintcache
+/.idea/
```
This should be in your global `.gitignore` file

https://docs.github.com/en/get-started/getting-started-with-git/ignoring-files#configuring-ignored-files-for-all-repositories-on-your-computer

```suggestion
```

BenoitAverty replied:
or if you don't want it global, in .git/info/exclude

donavon commented on packages/remix-react/scroll-restoration.tsx:
```diff
@@ -98,7 +98,9 @@ function useScrollRestoration() {
+      const regular = /^[0-9]\d+/;
+      const { hash } = location;
+      if (hash && !regular.test(hash.replace("#", ""))) {
         let el = document.querySelector(location.hash);
```
you've destructured `hash` so use it here

```suggestion
        let el = document.querySelector(hash);
```

chenc041 replied:
thank you  so much

tascord commented:
Hey folk, just wanted to add that some versions of different browsers may allow different characters as ID's, and thus as hashes.
We may just want to wrap things in a try/catch if that works with the convention of the project.
[A little info on id-s](https://mathiasbynens.be/notes/html5-id-class)

tascord commented:
Or, more specifically :: [here](https://html.spec.whatwg.org/multipage/dom.html#:~:text=There%20are%20no%20other%20restrictions%20on%20what%20form%20an%20ID%20can%20take%3B%20in%20particular%2C%20IDs%20can%20consist%20of%20just%20digits%2C%20start%20with%20a%20digit%2C%20start%20with%20an%20underscore%2C%20consist%20of%20just%20punctuation%2C%20etc.)

chenc041 commented:
> Hey folk, just wanted to add that some versions of different browsers may allow different characters as ID's, and thus as hashes. We may just want to wrap things in a try/catch if that works with the convention of the project. [A little info on id-s](https://mathiasbynens.be/notes/html5-id-class)

if someone use special id-s, i think we give warning in console and give some reference

tascord commented:
> > Hey folk, just wanted to add that some versions of different browsers may allow different characters as ID's, and thus as hashes. We may just want to wrap things in a try/catch if that works with the convention of the project. [A little info on id-s](https://mathiasbynens.be/notes/html5-id-class)
> 
> if someone use special id-s, i think we give warning in console and give some reference

Sounds good!

tascord commented:
Not sure how to commit changes myself, but I think this would be a good thing to put in place.
Also not sure if comms are written in the write voicing, so please edit to convention.
```ts
      if (location.hash) {
        if (!/^(#[A-z][A-z0-9-_:.]*)$/.test(location.hash)) {
          return console.warn(
            `The hash supplied, ${location.hash}, is not a valid HTML4 hash (https://www.w3.org/TR/html4/types.html#type-id). As HTML5 hashes aren-t supported in all browsers, scroll restoration will not work with them.`
          );
        }

        let el = document.querySelector(location.hash);
        if (el) {
          el.scrollIntoView();
          return;
        }
      }
```

chenc041 commented:
> Not sure how to commit changes myself, but I think this would be a good thing to put in place. Also not sure if comms are written in the write voicing, so please edit to convention.
> 
> ```ts
>       if (location.hash) {
>         if (!/^(#[A-z][A-z0-9-_:.]*)$/.test(location.hash)) {
>           return console.warn(
>             `The hash supplied, ${location.hash}, is not a valid HTML4 hash (https://www.w3.org/TR/html4/types.html#type-id). As HTML5 hashes aren't supported in all browsers, scroll restoration will not work with them.`
>           );
>         }
> 
>         let el = document.querySelector(location.hash);
>         if (el) {
>           el.scrollIntoView();
>           return;
>         }
>       }
> ```

nice 
I will edit i, thks 🤝

ryanflorence commented:
Oops, I thought I fixed this already! Thanks for the PR.

- Skip the regex check
- Remove the warning, some people might want numerical hashes for some reason and us warning them all the time would be annoying
- Use `document.getElementById(location.hash.slice(1))` because it doesn't blow up on numerical ids

And then let's get this merged!

chenc041 commented:
> Oops, I thought I fixed this already! Thanks for the PR.
> 
> * Skip the regex check
> * Remove the warning, some people might want numerical hashes for some reason and us warning them all the time would be annoying
> * Use `document.getElementById(location.hash.slice(1))` because it doesn't blow up on numerical ids
> 
> And then let's get this merged!

Hi, I have changed the code follow you suggest

dvargas92495 commented:
Hey, anything blocking this PR? Running into this issue on my end and would love to help get it merged

chenc041 commented:
> Hey, anything blocking this PR? Running into this issue on my end and would love to help get it merged

I think they have forgot...

chenc041 commented:
@ryanflorence hi ryan, If there is no problem, can you merge this PR as soon as possible?

na2hiro commented:
This fix enables use cases like using hash as URL params `#color=red` and it would be useful. Could you merge this? Thanks!

ryanflorence commented:
Thanks, sorry been a busy start of the year 😆

chenc041 commented:
> Thanks, sorry been a busy start of the year 😆

thanks, 😄
